pkg/controller/webhook/kubefedconfig/webhook.go

```diff
@@ -0,0 +1,124 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubefedconfig
+
+import (
+	"encoding/json"
+	"net/http"
+	"strings"
+	"sync"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1/validation"
+	"sigs.k8s.io/kubefed/pkg/controller/webhook"
+)
+
+const (
+	resourceName       = "KubeFedConfig"
+	resourcePluralName = "kubefedconfigs"
+)
+
+type KubeFedConfigValidationHook struct {
+	client dynamic.ResourceInterface
+
+	lock        sync.RWMutex
+	initialized bool
+}
+
+func (a *KubeFedConfigValidationHook) ValidatingResource() (plural schema.GroupVersionResource, singular string) {
+	return webhook.NewValidatingResource(resourcePluralName), strings.ToLower(resourceName)
+}
+
+func (a *KubeFedConfigValidationHook) Validate(admissionSpec *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
+	status := &admissionv1beta1.AdmissionResponse{}
+
+	if webhook.Allowed(admissionSpec, resourcePluralName) {
+		status.Allowed = true
+		return status
+	}
+
+	klog.V(4).Infof("Validating AdmissionRequest = %v", admissionSpec)
+
+	admittingObject := &v1beta1.KubeFedConfig{}
+	err := json.Unmarshal(admissionSpec.Object.Raw, admittingObject)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+	if !a.initialized {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusInternalServerError, Reason: metav1.StatusReasonInternalError,
+			Message: "not initialized",
+		}
+		return status
+	}
+
+	errs := validation.ValidateKubeFedConfig(admittingObject)
+	if len(errs) != 0 {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: errs.ToAggregate().Error(),
+		}
+		return status
+	}
+
+	status.Allowed = true
+	return status
+}
+
+func (a *KubeFedConfigValidationHook) Initialize(kubeClientConfig *rest.Config, stopCh <-chan struct{}) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	a.initialized = true
+
+	klog.V(2).Infof("KubeFedConfig validation webhook is initialized")
+
+	shallowClientConfigCopy := *kubeClientConfig
```

Comment:
What makes this a copy?  Is the caller making a copy before calling this function?

Reply:
For this one, I guess we need our own group version and it is not expected to change framework passed config argument.

cc @font

Reply:
> What makes this a copy? Is the caller making a copy before calling this function?

The caller creates a new in-cluster config, whose pertinent details (aside from `GroupVersion`, `APIPath`) are shared by all admission webhooks in this server, before invoking each webhook's `Initialize`.